packages/ra-ui-materialui/src/field/ArrayField.tsx

```diff
@@ -74,30 +75,35 @@ import { PublicFieldProps, InjectedFieldProps, fieldPropTypes } from './types';
  *
  * @see useListContext
  */
-export const ArrayField = memo<ArrayFieldProps>(props => {
+const ArrayFieldImpl = <
+    RecordType extends Record<string, unknown> = Record<string, any>
+>(
+    props: ArrayFieldProps<RecordType>
+) => {
     const { children, resource, source, perPage, sort, filter } = props;
     const record = useRecordContext(props);
-    const data = get(record, source, emptyArray) || emptyArray;
+    const data =
+        (get(record, source, emptyArray) as Record<string, any>[]) ||
+        emptyArray;
     const listContext = useList({ data, resource, perPage, sort, filter });
     return (
         <ListContextProvider value={listContext}>
             {children}
         </ListContextProvider>
     );
-});
-
-// @ts-ignore
-ArrayField.propTypes = {
-    ...fieldPropTypes,
 };
+ArrayFieldImpl.propTypes = fieldPropTypes;
```

Comment:
We used to clone the prop types before. There was probably a reason, I think you shouldn't change that.

Reply:
I can't think of a valid reason. We do the same in EmailField.

I think someone did not simplify the syntax when removing an old propType specific to the ArrayField that did not make sense in v4. See https://github.com/marmelab/react-admin/commit/f7d9f18505996a3af92f29b31d3bc1d104052dfe

Reply:
I vaguely remember having to clone proptypes for an obscure reason. Let's revert it if we have no reason to do it.

Reply:
What about the EmailField, UrlField and WrapperField then?

Reply:
Let's keep them as in master. I know, in master there is no consistency. Let's keep it like that. 

My opinion is that this PR changes quite a lot of stuff, and we should avoid all unnecessary changes to reduce any potential BC break.